### ISSUE TITLE
Wavetable Oscillator object

### ISSUE THREAD

**Issue by tap** [jamoma/jamoma2#5]
and maybe grain playback

**Comment by tap** [comment]
with @nwolek

**Comment by tap** [comment]
@nwolek This task, which is very similar to a grain playback object, seems like something that might best be approach with us doing it as a pair-programming project.  

Not sure what your schedule is like, but maybe sometime next week would work for tackling this one together?

**Comment by nwolek** [comment]
I agree we should would together on this. My schedule is pretty packed between now and Aug 21, as I finish this installation and start classes. I could Skype the morning of 10th but would not be able to follow up for a full 2 weeks after. 

During this busy stretch, I am going to be working on a interpolating Delay unit as time permits. This should prevent me from getting rusty. :) Started last night here: 
https://github.com/jamoma/jamoma2/commit/3a221b127bcee8c4f75672d447713a85f40be9c5

**Comment by tap** [comment]
Groovy!  Probably obvious, but in this case it is probably useful to use TTDelay as a reference/model.

**Comment by nwolek** [comment]
After reading about the topic some more, I think this object should be called "waveform" instead of "wavetable". Generator names should describe what they produce, not the technique used to generate them. The job of documenting the technique should be handled by the Doxygen description, not the name. 

Would you agree @tap @lossius ?

**Comment by tap** [comment]
How do we handle naming of additional different (non-wavetable) waveform generators?

**Comment by nwolek** [comment]
Would PeriodicWaveform be more specific?

My current thinking:
- PeriodicWaveform would have an internal `mSync` so that it is controlled by `mFrequency` and `mPhase` parameters. You could also select from common basic shapes like sine, cosine, sawtooth, triangle, etc. that are available in [JamomaGenerator.h](https://github.com/jamoma/jamoma2/blob/master/include/core/JamomaGenerator.h)
- Wavetable would require input from an external [Sync](https://github.com/jamoma/jamoma2/blob/master/include/objects/JamomaSync.h) to drive it with the phase.

In Max, these are two different modes of the `cycle~` object, but I think it would be cleaner for us to make them different classes in our library. Especially since one requires an input and the other does not. Willing to be convinced otherwise.

**Comment by tap** [comment]
I'm kind of just waving my hands about the naming thing.  On a certain level I guess I don't care what it is called and Wavetable makes it obvious to me what it is.  It could also be Oscillator.

Max's `cycle~` has 3 modes:
1. freq and phase are constants 
2. freq is a signal, phase is a constant
3. freq is a constant, phase is a signal

I don't have a clear vision on whether it should all be one class or not.  Clearly there would overlap (duplication?) between multiple classes. But it could also be easier to reason about the problem and develop as multiple classes, possibly merging them later or keeping them that way.

?

**Comment by nwolek** [comment]
I edited in numbers to your list for easier reference. Hope that is OK.

I guess what I have started working toward is item 1. I see item 3 as clearly a `Wavetable`, which is similar in function to the `wave~` object. But item 1 & 2 I am more likely to call an `Oscillator` (which admittedly is less typing than `PeriodicWaveform`).

I am inclined to build them separately, then see if we can merge them.

**Comment by nwolek** [comment]
Progress report on issue #5:
- The new padding functions (issue #82) resolve the tableSize concern.
- I now think that `Sync` members `gain` and `offset` should transition to be expressed as minimum and maximum. This would reduce confusion.
- Need to implement waveshape options.
- Need to implement interpolation options.

**Comment by nwolek** [comment]
@tap I need your help on these two things. I made 2 or 3 attempts to implement yesterday, but was not happy with the results each time.
- Need to implement waveshape options.
- Need to implement interpolation options.

Maybe you could email me to set up a time early next week?

**Comment by nwolek** [comment]
On hold pending progress on Issue #89
